source/includes/fundamentals/code-snippets/Monitoring.java

```diff
@@ -0,0 +1,151 @@
+        collection.find().first();
+        mongoClient.close();
+        // end monitor-command-example
+    }
+
+    private void monitorClusterEvent() {
+        // start monitor-cluster-example
+        IsWriteable clusterListener = new IsWriteable();
+        MongoClientSettings settings =
+                MongoClientSettings.builder()
+                        .applyConnectionString(URI)
+                        .applyToClusterSettings(builder ->
+                                builder.addClusterListener(clusterListener))
+                        .build();
+        MongoClient mongoClient = MongoClients.create(settings);
+        MongoDatabase database = mongoClient.getDatabase(DATABASE);
+        MongoCollection<Document> collection = database.getCollection(COLLECTION);
+        // Run a command to trigger a ClusterDescriptionChangedEvent event
+        collection.find().first();
+        // end monitor-cluster-example
+        mongoClient.close();
+    }
+
+    private void monitorConnectionPoolEvent() {
+        // start monitor-cp-example
+        ConnectionPoolLibrarian cpListener = new ConnectionPoolLibrarian();
+        MongoClientSettings settings =
+                MongoClientSettings.builder()
+                        .applyConnectionString(URI)
+                        .applyToConnectionPoolSettings(builder ->
+                                builder.addConnectionPoolListener(cpListener))
+                        .build();
+        MongoClient mongoClient = MongoClients.create(settings);
+        MongoDatabase database = mongoClient.getDatabase(DATABASE);
+        MongoCollection<Document> collection = database.getCollection(COLLECTION);
+        // Run a command to trigger connection pool events
+        collection.find().first();
+        // end monitor-cp-example
+        /* We do not close this connection in order to prevent the driver from requesting two connections, giving
+        the example unintuitive output. Uncomment the following line of code and see how it effects this example*/
+        //mongoClient.close();
+    }
+
+}
+
+// start command-listener-impl
+class CommandCounter implements CommandListener {
+
+    private Map<String, Integer> commands = new HashMap<String, Integer>();
+
+    @Override
+    public void commandSucceeded(final CommandSucceededEvent event) {
+        String commandName = event.getCommandName();
+        int count = commands.containsKey(commandName) ? commands.get(commandName) : 0;
+        commands.put(commandName, count + 1);
+        System.out.println(commands.toString());
+    }
+
+    @Override
+    public void commandFailed(final CommandFailedEvent event) {
+        System.out.println(String.format("Failed execution of command '%s' with id %s",
+                event.getCommandName(),
+                event.getRequestId()));
+    }
+}
+// end command-listener-impl
+
+// start cluster-listener-impl
+class IsWriteable implements ClusterListener {
+
+    private boolean isWritable;
```

Comment:
`private volatile boolean`.

Otherwise there is a race condition.

Reply:
Thanks for the note. Fix: https://github.com/mongodb/docs-java/pull/88/commits/841ecea2470d2b560f92be1793d4e88080183164

Question for my own edification:

I am not an expert of usage of `volatile`, but does the usage of `isWritable` as an invariant in the conditional disqualify it from usage of `volatile` (and thus require `synchronized` in the method signature)? 

[Old Java Concurrency Volatile Discussion:](http://web.archive.org/web/20210221170926/https://www.ibm.com/developerworks/java/library/j-jtp06197/)

Conditions for usage of volatile:

>  Writes to the variable do not depend on its current value.
The variable does not participate in invariants with other variables.

In researching volatile, a theme I am running into is "don't use a volatile variable if it's previous state will determine something about its future state". In this specific code snippet, I am having trouble coming up with a scenario that would actually create problems (if `isWritable` is written to by another thread in between reading in conditional and writing new boolean value I don't think this will create problems in the code), but based off my current understanding of conventions/patterns around `volatile` usage I am not sure which is more appropriate in this scenario.

Reply:
Thanks for asking, as it forced me to look more deeply at this, and it turns out I am wrong about it.  I thought that the driver guaranteed there are no concurrent calls to this method, but that's not actually true: the driver calls this method without holding any exclusive lock.  So I think it is more appropriate to make the method synchronized instead of making the field volatile.

Reply:
Ok, sounds good. Thanks for looking into this and helping me better understand usage of `volatile` and `synchronized`. Fix: https://github.com/mongodb/docs-java/pull/88/commits/7643f1a15acc72947e41a62f6a4328ccb009ebf2